cmd/cstor-pool-mgmt/volumereplica/volumereplica.go

```diff
@@ -699,24 +700,127 @@ func GetAndUpdateSnapshotInfo(
-//}
+// getAndAddPendingSnapshotList get the snapshot information from peer replicas and
+// add under pending snapshot list
+// NOTE: Below function will delete the snapshot under pending snapshots if doesn't exists
+// on peer replicas
+func getAndAddPendingSnapshotList(
+	clientset clientset.Interface, cvr *apis.CStorVolumeReplica) error {
+	peerCVRList, err := getPeerReplicas(clientset, cvr)
+	if err != nil {
+		return errors.Wrapf(err, "failed to get peer CVRs of volume replica %s", cvr.Name)
+	}
+
+	peerSnapshotList := getPeerSnapshotInfoList(peerCVRList)
+	if cvr.Status.PendingSnapshots == nil {
+		cvr.Status.PendingSnapshots = map[string]apis.CStorSnapshotInfo{}
+	}
+
+	// Delete the pending snapshot that doesn't exist in peer snapshot list
+	for snapName, _ := range cvr.Status.PendingSnapshots {
+		if _, ok := peerSnapshotList[snapName]; !ok {
+			delete(cvr.Status.PendingSnapshots, snapName)
+		}
+	}
+
+	// Add peer snapshots if doesn't exist on Snapshots and PendingSnapshots
+	// of current CVR
+	for snapName, snapInfo := range peerSnapshotList {
+		if _, ok := cvr.Status.Snapshots[snapName]; !ok {
+			if _, ok := cvr.Status.PendingSnapshots[snapName]; !ok {
+				cvr.Status.PendingSnapshots[snapName] = snapInfo
+			}
+		}
+	}
+
+	return nil
+}
+
+// getPeerReplicas returns list of peer replicas of volume
+func getPeerReplicas(
+	clientset clientset.Interface,
+	cvr *apis.CStorVolumeReplica) (*apis.CStorVolumeReplicaList, error) {
+	volName := cvr.GetLabels()[string(apis.PersistentVolumeCPK)]
+	peerCVRList := &apis.CStorVolumeReplicaList{}
+	cvrList, err := clientset.OpenebsV1alpha1().
+		CStorVolumeReplicas(cvr.Namespace).
+		List(metav1.ListOptions{LabelSelector: string(apis.PersistentVolumeCPK) + "=" + volName})
+	if err != nil {
+		return nil, err
+	}
+	for _, obj := range cvrList.Items {
+		if obj.Name != cvr.Name {
+			peerCVRList.Items = append(peerCVRList.Items, obj)
+		}
+	}
+	return peerCVRList, nil
+}
+
+// getPeerSnapshotInfoList returns the map of snapshot name and snapshot info
+// If any healthy replica exist in peer replica it will return Status.Snapshots
+// else iterate over all the degraded replicas and get snapshot list
+func getPeerSnapshotInfoList(
+	peerCVRList *apis.CStorVolumeReplicaList) map[string]apis.CStorSnapshotInfo {
+
+	// TODO: Get Opinion From Review Comments
+	// NOTE: There are possibilites to have stale phase
+	healthyReplica := getHealthyReplicaFromPeerReplicas(peerCVRList)
+	if healthyReplica != nil {
+		// Since Updating the status of replica and snapshot list is atomic
+		// update safe to return status.snapshots
+		return healthyReplica.Status.Snapshots
+	}
+
+	snapshotInfoList := map[string]apis.CStorSnapshotInfo{}
+	for _, cvrObj := range peerCVRList.Items {
+		// No need to get snapshot information from Offline,
+		// NewReplicaDegraded and Recreate CVR becasue they might
+		// consist stale information
+		if cvrObj.Status.Phase == apis.CVRStatusDegraded {
```

Comment:
why only degraded? Other replica can be in rebuilding phase also.

Reply:
You are correct this check doesn't make sense I will remove it.